batchproducer/batchproducer.go

```diff
@@ -260,7 +282,15 @@ func (b *batchProducer) sendBatch() {
 		b.consecutiveErrors++
 		b.currentStat.KinesisErrorsSinceLastStat++
 		b.logger.Printf("Error occurred when sending PutRecords request to Kinesis stream %v: %v", b.streamName, err)
-		b.returnRecordsToBuffer(records)
+
+		if b.consecutiveErrors >= 5 && b.isBufferFullOrNearlyFull() {
+			// In order to prevent Add from hanging indefinitely, we start dropping records
+			b.logger.Printf("DROPPING %v records because buffer is full or nearly full and there have been %v consecutive errors from Kinesis", len(records), b.consecutiveErrors)
+		} else {
+			b.logger.Printf("Returning %v records to buffer (%v consecutive errors)", len(records), b.consecutiveErrors)
```

Comment:
This is not a TH go project; this is an open-source project by send grid and this repo is a fork. I used the existing conventions of the repo because I plan to contribute this addition back to the original project.